src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -60,7 +60,7 @@ public TextDocumentState(SolutionServices solutionServices, DocumentInfo info, L
                info.DocumentServiceProvider,
                info.Attributes,
                textAndVersionSource: info.TextLoader != null
-                ? CreateRecoverableText(info.TextLoader, solutionServices)
+                ? CreateTextFromLoader(info.TextLoader, PreservationMode.PreserveValue, solutionServices)
```

Comment:
all the logic and helpers unified on one helper `CreateTextFromLoader`

Reply:
the AlwaysHoldStrongly flag was needed so that creating a fresh document out of one of these loaders would know the semantics needed to hold it by.  Previously, this would *always* make a recoerabletext locally which would dump the contents to a MMF in teh server itself (pointless since it's already in an MMF file in the  host that we're sharing).

Reply:
i looked into having some sort of Mode flow in here, but it was very viral, and i viewed it as more reasonable to just allow the Loader to have a say in if it should be held strongly, or if a RecoverableTexxt should be used.